### DISCUSSION TITLE
Changing icon photo

### DISCUSSION THREAD

**Discussion by Moosems** [pyinstaller/pyinstaller#6934] [Help]
How can I allow my Mac app to change its icon photo to another file after it is bundled and windowed? I know that while the app is running it can be changed by the GUI toolkit but for the actual bundle I am unsure how to change it

**Comment by rokm** [comment]
The app bundles are supposed to be immutable, so I doubt the OS allows the application to modify its .app bundle. I.e., if the bundle is unsigned, it is likely moved and ran from a temporary read-only location, and if it is signed, (self-)modifications will void the signature anyway. 

But FWIW, the bundle icon is determined by the `CFBundleIconFile` key in `name.app/Contents/Info.plist` and the referenced icon file is located in `name.app/Contents/Resources`.

**Reply by Moosems** [reply]
What if I used the same file name but changed the icon in `name.app/Contents/Resources`? I know that iOS apps can change their icon and when you look at their files they're just Mac apps in some more folders

**Reply by Moosems** [reply]
> The app bundles are supposed to be immutable

Then how can I allow internal files to be saved or edited? For example, you can change the theme on discord and it saves it to a file but doesn't have issues. Just wondering so I can be a more effective developer

**Reply by rokm** [reply]
I doubt it would work from within application (at runtime) without being prevented by translocation or invalidating the signature. 

Perhaps there is a mechanism that allows icon switching like what you describe for iOS apps, but as I have very little vested interest in macOS and macOS apps, I am not aware of it.

**Reply by rokm** [reply]
> > The app bundles are supposed to be immutable
> 
> Then how can I allow internal files to be saved or edited? 

You don't. Even if this wasn't a macOS bundle with its own unique requirements, the frozen application may be installed in a read-only location, or used by several different users, etc., so the frozen application should be as immutable as possible.

>For example, you can change the theme on discord and it saves it to a file but doesn't have issues. Just wondering so I can be a more effective developer

Different OSes have different mechanisms for storing persistent settings and state, like registry keys on Windows, or config files in user's home directory on linux (e.g., ~/.config/appname). According to [SO](https://stackoverflow.com/questions/410013/where-do-osx-applications-typically-store-user-configuration-data), on macOS that would be `/Users/username/Library/Preferences` or `/Users/username/Library/Application Support`. This might also be handled by the toolkit (i.e., Qt provides abstraction for dealing with persistent settings on different OSes), or you could use something like [`appdirs`](https://pypi.org/project/appdirs/).
